docs/bsp.md

```diff
@@ -717,6 +717,10 @@ trait SourceItem {
     * directory belongs to this source item.
     */
   def uri: Uri
+  
+  /** Specifies whether uri denotes a directory or regular file. 
+    * If this is not set, and the uri does not point to an existing file, it should be treated as directory by default. */
+  def isDirectory: Option[Boolean]
```

Comment:
Can we make this a `kind: Int` field with possible values `FILE` or `DIRECTORY`? A boolean `isX` option is often an enum screaming to get out!

Reply:
I'm not sure that makes sense in this case. Do we have anything else these paths could potentially be?

Reply:
I think `if (!item.isDirectory)` is less clear than `if (item.kind == FILE)`. Also, with kind we could drop the `generated: Boolean` option and instead use 

```
enum Kind { FILE, DIRECTORY, GENERATED_FILE, GENERATED_DIRECTORY }
```

Reply:
Agree on clarity, though boolean encoding is simpler.

generated and directory are orthogonal, so I don't think it makes sense to combine them into a single enum (though that is how it's done within IntelliJ)

Overall I'm not very emotional on this, so let's see what @jvican says

Reply:
I prefer the enum encoding over the boolean to keep the API open and avoid API gotchas in the future. Otherwise, what `isDirectory` meant in the past could be broken with the introduction of new types in the future.

Reply:
Okay, there's several ways we can do it:
* a field `kind` which is one of `FILE` or `DIRECTORY`, and keep `generated` Boolean.
* a field `kind` which is one of `FILE`, `DIRECTORY`, `GENERATED_FILE`, `GENERATED_DIRECTORY` and replaces `generated`. This has the drawback that if we add more kinds, expecially orthogonal kinds, it would bloat the possible enum values.
* separate field for `kind` and `generated`, both containing enums.
* a field `kinds` OR `tags` which can contain 0 or more of `FILE`, `DIRECTORY`, `GENERATED`, which replaces both booleans. Files would default to non-generated directories.

The tag approach is most flexible, but makes the API less explicit in declaring that a file/dir needs to be explicitly declared.

Reply:
I like Olafur's suggestion to define it like `enum Kind { FILE, DIRECTORY, GENERATED_FILE, GENERATED_DIRECTORY }`

Reply:
Why do you prefer it? It doesn't expand well when adding more attributes. Anyway, I've pushed a version using the `kinds` list for your consideration.

Reply:
`List[Kind]` allows representing invalid values like `[File,Directory,File]`. I like `enum Kind { FILE, DIRECTORY, GENERATED_FILE, GENERATED_DIRECTORY }` since it represents only valid values. I am OK with doubling the number of enum values for every new attribute, I don't expect we'll add too many attributes in the future.

Reply:
Okay, let's go with that then.